src/binary-encoding.adoc

```diff
@@ -0,0 +1,1600 @@
+0 1 0 0 0 0 1 1  0 0 0 0 1 0 0 1
+└──┬──┘ └──┬──┘  └──────┬──────┘
+   │       │            └──────────── VarUInt containing the 7 most
+   │       └── 4 least significant    significant bits of the macro
+opcode high    bits of the macro      address
+nibble 4       address
+
+Magnitude bits: 0000100_0011
+Decoded value : 67
+Biased value  : 131
+----
+
+==== Example encoding of an invocation of the macro at address `1211`
+[source]
+----
+
+                               ┌─── The address VarUInt ends in a `1`,
+                               │    no more VarUInt bytes follow.
+                               │
+0 1 0 0 1 0 1 1  1 0 0 0 1 1 1 1
+└──┬──┘ └──┬──┘  └──────┬──────┘
+   │       │            └──────────── VarUInt containing the 7 most
+   │       └── 4 least significant    significant bits of the macro
+opcode high    bits of the macro      address
+nibble 4       address
+
+Magnitude bits: 1000111_1011
+Decoded value : 1,147
+Biased value  : 1,211
+----
+
+==== Example encoding of an invocation of the macro at address `71376`
+[source]
+----
+
+                              ┌─── The address VarUInt ends in `10`; the trailing
+                              │    zero indicates that one more VarUInt byte follows.
+                             ┌┴┐
+0 1 0 0 0 0 0 0  1 0 1 0 0 1 1 0  0 1 0 0 0 1 0 1
+└──┬──┘ └──┬──┘  └──────┬──────┘  └──────┬──────┘
+   │       │            │                └──────────── the 8 most significant bits
+   │       │            │                              of the macro address
+   │       │            │
+   │       │            └──────────── VarUInt containing the next 7 most
+   │       └── 4 least significant    significant bits of the macro
+opcode high    bits of the macro      address
+nibble 4       address
+
+Magnitude bits: 01000101_101001_0000
+Decoded value : 71,312
+Biased value  : 71,376
+----
+
+NOTE: From this point on in the document, example encodings are given in hexidecimal notation.
+
+[[booleans]]
+== Booleans
+
+`0x5E` represents boolean `true`, while `0x5F` represents boolean `false`.
+
+// XXX: Structurally, these examples should be at depth 3. However, all other example sections in the doc
+//      are at heading depth 4, which keeps them out of the TOC. These are also at depth 4 for consistency.
+
+==== Example encoding of `true`
+[source]
+----
+5e
+----
+
+==== Example encoding of `false`
+[source]
+----
+5f
+----
+
+[[numbers]]
+== Numbers
+
+[[integers]]
+=== Integers
+
+Opcodes in the range `0x50` to `0x58` represent an integer. The opcode is followed by a <<fixedint, `FixedInt`>> that
+represents the integer value. The low nibble of the opcode (`0x_0` to `0x_8`) indicates the size of the `FixedInt`.
+Opcode `0x50` represents integer `0`; no more bytes follow.
+
+Integers that require more than 8 bytes are encoded using the variable-length integer opcode `0xF5`,
+followed by a
+<<varuint, VarUInt>> indicating how many bytes of representation data follow.
```

Comment:
Should there be an example for that below?

Reply:
There's an example of variable-length encoding, it's just being used on a value that's only a few bytes long.

```
==== Example variable-length encoding of `-944`
[source]
----
┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
│   ┌─── VarUInt 2; a 2-byte FixedInt follows
│   │
F5 05 50 fc
      └─┬─┘
   FixedInt -944
----